Embind: Make subclasses inherit class functions

JeanChristopheMorinPerso opened the pull request:
Attempt to fix #18722.

ES6 style subclasses automatically inherit their parent's static functions. ES5 style classes don't. But I thought it would be nice if embind classes behaved more like ES6 classes.

The approach I took is to keep track of child classes in `<BaseClass>.__derivedClasses`, and then apply the class functions on child classes in `_embind_register_class_class_function`.

I initially tried to apply all class functions in `_embind_register_class`, but the problem with that is that the functions on the base class are not yet resolved (it's set to the `unboundTypesHandler` function). So the child class would end up with a class function that is never resolved (which would raise `Cannot call <class> due to unbound types ...`).

I don't know if there is a better way to achieve that. I'm open to suggestions!

JeanChristopheMorinPerso commented on src/embind/embind.js:
```diff
@@ -2371,6 +2380,15 @@ var LibraryEmbind = {
+        if (classType.registeredClass.__derivedClasses) {
+          for (const derivedClass of classType.registeredClass.__derivedClasses) {
+            if (!derivedClass.constructor.hasOwnProperty(methodName)) {
+              derivedClass.constructor[methodName] = func;
```
This doesn't consider overloads... I'm guessing I would need to?

brendandahl replied:
To match how C++ would work we probably should, but I don't think it's needed right now. It also doesn't look like overloading completely works for regular class methods right now anyway.

Adding a todo here would be good though.

JeanChristopheMorinPerso replied:
I added a todo.

brendandahl commented on src/embind/embind.js:
```diff
@@ -2063,6 +2063,15 @@ var LibraryEmbind = {
                                                   downcast);
 
+        if (registeredClass.baseClass) {
+          // Keep track of class hierarchy. USed to allow sub-classes to inherit class functions.
```
nit: `Used`

JeanChristopheMorinPerso commented:
Hello, I addressed the comments. I was wondering if someone could potentially review this PR please? I'd really love to see this land on the main branch :)

kripken commented:
@brendandahl do you have time to look at this?

JeanChristopheMorinPerso commented:
Done @brendandahl !

JeanChristopheMorinPerso commented:
Some tests failed but it doesn't seem related to my changes...

sbc100 commented:
Can you put "embind" in the PR description somewhere so its clear what this change is about?

JeanChristopheMorinPerso commented:
@sbc100 I added embind in the PR title. Does that work?
